### PR TITLE
Support { } in delctypes embedded within templates

### PR DESCRIPTION
**Pull request by parkerjs**
- Modified the check_template() function to account for braces that may be
  embedded within a decltype statement that itself is within a template
  argument list. Without this modification, check_template() incorrectly
  interprets the angle brackets within a template as comparisons

- Moved existing code already used to detect braced initializer statements
  into a separate function so that check_template() may leverage this
  detection code as well

- Added new flag 'PCF_IN_DECLTYPE' to indicate chunk within cpp decltype
  statement

- Made modifications to mark chunks belonging to decltype statements

- Updated cli test scripts and expected output to reflect additions to
  pcf_flag_e bitfields.

### CODE REVIEW

**Comment by parkerjs** [comment]
I tried to change as little as possible to add support for the referenced subject. Hopefully I followed established protocols/precedents, but feel free to let me know if I should do things differently.

File: src/combine.cpp [hunk 1]
```diff
@@ -1155,72 +1155,28 @@ void do_symbol_check(chunk_t *prev, chunk_t *pc, chunk_t *next)
...
-         log_pcf_flags(LFCNR, pc->flags);
-         auto brace_open = chunk_get_next_ncnl(pc);
+         tmp = chunk_get_next_ncnl(brace_close);
 
-         if (  chunk_is_token(brace_open, CT_BRACE_OPEN)
-            && (  get_chunk_parent_type(brace_open) == CT_NONE
-               || get_chunk_parent_type(brace_open) == CT_ASSIGN
-               || get_chunk_parent_type(brace_open) == CT_RETURN
-               || get_chunk_parent_type(brace_open) == CT_BRACED_INIT_LIST))
+         if (tmp)
```
Comment (gmaurel) [inline src/combine.cpp:1169]: Please use:
if (tmp != nullptr)

File: src/combine.cpp [hunk 2]
```diff
@@ -3832,3 +3788,65 @@ static void handle_java_assert(chunk_t *pc)
       }
    }
 }
+
+
+/**
+ * Detect a cpp braced init list
+ */
+bool detect_cpp_braced_init_list(chunk_t *pc, chunk_t *next)
```
Comment (gmaurel) [inline src/combine.cpp:3796]: Please put the function in extra files (.cpp, .h)
Use an #include ...
Modify the file CMakeLists.txt, to take care

File: src/combine.h [hunk 3]
```diff
@@ -74,4 +74,10 @@ chunk_t *set_paren_parent(chunk_t *start, c_token_t parent);
 void do_symbol_check(chunk_t *prev, chunk_t *pc, chunk_t *next);
 
 
+/**
+ * Detect a cpp braced init list
+ */
+bool detect_cpp_braced_init_list(chunk_t *pc, chunk_t *next);
```
Comment (gmaurel) [inline src/combine.h:80]: same as above

File: src/tokenize_cleanup.cpp [hunk 4]
```diff
@@ -92,6 +92,12 @@ static void mark_selectors_in_property_with_open_paren(chunk_t *open_paren);
 static void mark_attributes_in_property_with_open_paren(chunk_t *open_paren);
 
 
+/**
+ * Mark all chunks within the scope of a decltype() statement
+ */
+static void mark_cpp_decltype(chunk_t *pc);
```
Comment (gmaurel) [inline src/tokenize_cleanup.cpp:98]: same as above

File: src/tokenize_cleanup.cpp [hunk 5]
```diff
@@ -1189,8 +1204,8 @@ static void check_template(chunk_t *start, bool in_type_cast)
    {
       /*
        * We may have something like "a< ... >", which is a template where
-       * '...' may consist of anything except braces {}, a semicolon, and
-       * unbalanced parens.
+       * '...' may consist of anything except braces {} (unless perhaps
+       * embedded within a decltype), a semicolon, and unbalanced parens.
```
Comment (gmaurel) [inline src/tokenize_cleanup.cpp:1208]: put some more example(s) here

File: src/tokenize_cleanup.cpp [hunk 6]
```diff
@@ -1219,13 +1234,37 @@ static void check_template(chunk_t *start, bool in_type_cast)
...
       {
          if (  (chunk_is_token(pc, CT_SEMICOLON) && hit_semicolon)
-            || chunk_is_token(pc, CT_BRACE_OPEN)
-            || chunk_is_token(pc, CT_BRACE_CLOSE)
             || chunk_is_token(pc, CT_SQUARE_CLOSE))
          {
             break;
          }
 
+         if (pc->type == CT_DECLTYPE)
```
Comment (gmaurel) [inline src/tokenize_cleanup.cpp:1242]: Use the function chunk_is_token(pc, CT_DECLTYPE)

File: src/tokenize_cleanup.cpp [hunk 7]
```diff
@@ -1219,13 +1234,37 @@ static void check_template(chunk_t *start, bool in_type_cast)
...
+         {
+            mark_cpp_decltype(pc);
+         }
+
+         if (chunk_is_token(pc, CT_BRACE_OPEN))
```
Comment (gmaurel) [inline src/tokenize_cleanup.cpp:1247]: ... such as here

File: src/tokenize_cleanup.cpp [hunk 8]
```diff
@@ -1567,3 +1623,30 @@ static void mark_attributes_in_property_with_open_paren(chunk_t *open_paren)
       tmp = tmp->next;
    }
 }
+
+
+void mark_cpp_decltype(chunk_t *pc)
```
Comment (gmaurel) [inline src/tokenize_cleanup.cpp:1628]: same as above

File: tests/cli/output/p.txt [hunk 9]
```diff
@@ -8,72 +8,72 @@ newlines                        = crlf
 # language                      = CPP
```
Comment (gmaurel) [inline tests/cli/output/p.txt:8]: The file p.txt should NOT be part of the PR

File: src/combine.cpp (hunk 2 above)
Comment (parkerjs) [reply src/combine.cpp:3796]: What shall I name the new files? Any suggestions? Thanks

File: src/combine.h (hunk 3 above)
Comment (parkerjs) [reply src/combine.h:80]: Same as previous remark.

File: src/tokenize_cleanup.cpp (hunk 4 above)
Comment (parkerjs) [reply src/tokenize_cleanup.cpp:98]: Same as previous remark. I ask because you probably have a better idea as to what the new files should be called.

File: tests/cli/output/p.txt (hunk 9 above)
Comment (parkerjs) [reply tests/cli/output/p.txt:8]: Ok, so I updated this file because I added PCF_IN_DECLTYPE to the bitfields in pcf_flag_e. If I don't update this, I fail the CLI options test. Please advise.

File: src/combine.cpp (hunk 2 above)
Comment (parkerjs) [reply src/combine.cpp:3796]: I moved this function into a new file, 'flag_braced_init_list.cpp'

File: src/tokenize_cleanup.cpp (hunk 4 above)
Comment (parkerjs) [reply src/tokenize_cleanup.cpp:98]: I moved this function into a new file, 'flag_decltype.cpp'

File: tests/cli/output/p.txt (hunk 9 above)
Comment (guy-maurel) [reply tests/cli/output/p.txt:8]: run uncrustify/tests/cli/test_cli_options.py
to see what has changed, and review the output/xyz.txt file

Comment (parkerjs) [reply tests/cli/output/p.txt:8]: I've reviewed the results of the diff, and I think the reason for which I updated the p.txt output file still stands. I've inserted the PCF_IN_DECLTYPE flag at bit 16 (17th LSB), and as such the "Flag" field must be updated to reflect this. For example, if you take line 11 of the existing p.txt output file, you'll see that the value under the Flag column is 1000e0001; however, that value is before I inserted the new flag. With the addition of PCF_IN_DECLTYPE at the 16th bit position, the upper 48 bits must be left-shifted by one (the most significant bit is discarded, but it's not used anyway). This means that 1000e0001 must be updated to the value 2001c0001 for line 11.
```
           1    0    0    0    e    0    0    0    1
Before: 0001 0000 0000 0000 1110 0000 0000 0000 0001
           2    0    0    1    c    0    0    0    1
After:  0010 0000 0000 0001 1100 0000 0000 0000 0001
                               ^
                               insertion here at 16th bit position (17th LSB)
```

I've updated the Flag fields where applicable - some Flag fields did not change since no bits higher than the 16th LSB were set in the first place. I hope this makes sense.
